### ISSUE TITLE
parallel-workload: DDLs can deadlock

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#28400]
### What version of Materialize are you using?

8dc2c2b375

### What is the issue?

Seen in https://buildkite.com/materialize/nightly/builds/8674#0190d511-ac1a-4172-a4ce-ec921658f002, nothing was running except these two DDLs at the time and it still failed:
```
2024-07-21 12:09:57 UTC	worker_0 still running: DROP MATERIALIZED VIEW "db-pw-1721561907-0"."s-0"."v-65";
2024-07-21 12:09:57 UTC	worker_1 still running: CREATE ROLE role86;
2024-07-21 12:09:57 UTC	Threads have not stopped within 5 minutes, exiting hard
```
They've been running for half of the test already, so definitely a deadlock. The test does nothing except run DDLs.
[parallel-workload-queries.log](https://github.com/user-attachments/files/16335546/parallel-workload-queries.log)
[services.log.zip](https://github.com/user-attachments/files/16335550/services.log.zip)

**Comment by benesch** [comment]
Marking as a release blocker. I assume this is a result of https://github.com/MaterializeInc/materialize/pull/28128.

**Comment by maddyblue** [comment]
Investigating.

**Comment by maddyblue** [comment]
Is this possibly the same bug as in https://github.com/MaterializeInc/materialize/issues/28182 which was present before #28128 merged?

**Comment by maddyblue** [comment]
I added some debug code on main dealing with the serialized ddl additions and have been running the above command (though without the seed after the first few times) and so far have been unable to reproduce this.

**Comment by def-** [comment]
It's possible, but since I don't know the root cause for #28182 difficult to say. Renaming seemed to always cause it.

**Comment by maddyblue** [comment]
Nevermind, got it to repro with 10 threads.

**Comment by maddyblue** [comment]
Here's a log with some added printlns:

```
conn 132414 got lock for stmt CREATE ROLE role185
conn 305395 enqueing for stmt DROP SOURCE postgres_source5
conn 132414 clearing, serialized_ddl len: 1
handle_deferred_statement: 1
got deferred: Statement { stmt: DropObjects(DropObjectsStatement { object_type: Source, if_exists: false, names: [Item(UnresolvedItemName([Ident("postgres_source5")]))], cascade: false }), params: Params { datums: Row{[]}, types: [] } }
conn 305395 got lock for stmt DROP SOURCE postgres_source5
conn 412777 enqueing for stmt DROP SOURCE postgres_source8
conn 305395 enqueing for stmt DROP MATERIALIZED VIEW "db-pw-1721680928-0"."s-0"."v-13"
```

Connection 305395 got a lock, and then soon after is also enqueing another statement, but there's no println that is clearing conn 305395. This suggests that the txn cleanup function is not getting called in all cases, and the design assumes that it is.

**Comment by maddyblue** [comment]
Results of the investigation from today:

The workload test uses the http endpoint, so there's at least a small chance the bug is there. One theory is that the logic around https://github.com/MaterializeInc/materialize/blob/e944d03dab75b25f947308c78debb6f9f1cfbcfe/src/environmentd/src/http/sql.rs#L1066 is incorrect. I observed an error returned by the `let err = SqlResult::err(client, err);` line: (`Err { error: SqlError { message: "role u24 was concurrently dropped", code: "42501", detail: Some("Please disconnect and re-connect with a valid role."), hint: None, position: None }, notices: [] }`). But the logic in that loop continues on, even if there's an error, with the next statement group. There's no comments there explaining why it's correct to continue in the face of error, and I'm not sure if it's right or not.

I did see an unexpected Command::Commit in the logs. The only other place that command is sent is in the execute_single_statement task. I've added some additional logging to see if that's the one being fired and am waiting for it to trigger. (Reproducing the errors takes ~30m each time, so this is a slow process.)

Summary of two current places to investigate:
- execute_single_statement is maybe not handling errors correctly and failing to run the session terminate code
- the http endpoint that executes statement groups is possibly not handling errors correctly

A bit of evidence suggests that this problem could be occurring when the DDL parallel workload sends a DROP ROLE or REVOKE while another session is running causing that session to error, and one of our more infrequent logic paths does the wrong thing, doesn't clear the transaction or session, causing the lock to not be released, and the DDL is the orphaned waiting forever.

One workaround is that when trying to acquire the lock, before enqueuing the session, ensure that active_conns doesn't have the lock already (this is the deadlock). If it does, immediately return an error to the client, and terminate the connection (because it's in some bad state). This should be medium straightforward to do but leave the underlying bug in place, but is at least something.

**Comment by def-** [comment]
> The workload test uses the http endpoint, so there's at least a small chance the bug is there.

This should run parallel-workload without http/websocket, I'll try if the DDLs still hang:
```diff
diff --git a/misc/python/materialize/parallel_workload/executor.py b/misc/python/materialize/parallel_workload/executor.py
index 7be7ba2c39..771121e4be 100644
--- a/misc/python/materialize/parallel_workload/executor.py
+++ b/misc/python/materialize/parallel_workload/executor.py
@@ -69,7 +69,7 @@ class Executor:
         self.rollback_next = True
         self.last_log = ""
         self.action_run_since_last_commit_rollback = False
-        self.use_ws = self.rng.choice([True, False]) if self.ws else False
+        self.use_ws = False

     def set_isolation(self, level: str) -> None:
         self.execute(f"SET TRANSACTION_ISOLATION TO '{level}'")
@@ -125,14 +125,12 @@ class Executor:
         http: Http = Http.NO,
         fetch: bool = False,
     ) -> None:
-        is_http = (
-            http == Http.RANDOM and self.rng.choice([True, False])
-        ) or http == Http.YES
+        is_http = False
         if explainable and self.rng.choice([True, False]):
             query = f"EXPLAIN {query}"
         query += ";"
         extra_info_str = f" ({extra_info})" if extra_info else ""
-        use_ws = self.use_ws and http != Http.NO
+        use_ws = False
         http_str = " [HTTP]" if is_http else " [WS]" if use_ws and self.ws else ""
         self.log(f"{query}{extra_info_str}{http_str}")
         if not is_http:
```

**Comment by def-** [comment]
Locally I haven't seen it without http/ws.

But the same failure seems to also have happened in other DDL workloads (with other actions): https://buildkite.com/materialize/nightly/builds/8728#0190df75-34bc-4667-a3a9-cc3945e6b2c6

**Comment by maddyblue** [comment]
Current best guess is that this bug existed before serialized ddl, but that feature exposed it more loudly by doing a worse thing (preventing all DDL) when an assumption was not met (that sessions correctly end their transactions). Also likely this is the bug causing the various hung sessions we've seen somewhat rarely in various issues (I'm unable to find those).

**Comment by maddyblue** [comment]
https://github.com/MaterializeInc/materialize/pull/28506 is a workaround that will leave this issue open but would remove its release-blocker status.

**Comment by maddyblue** [comment]
I may have found it: http and pgwire uses of the AdapterClient send `Command::Commit` messages which get turned into `Plan::CommitTransaction` and sent off `sequence_plan` for processing. There, before anything is executed, `check_plan` is run: https://github.com/MaterializeInc/materialize/blob/bb56335708629a637962bfc745794b65546d3c5a/src/adapter/src/coord/sequencer.rs#L124

If that returns an error the ctx is retired before any execution is done. I think we need to change things so that `Command::Commit` will always run its cleanup? Or `check_plan` needs to be taught that if a role was dropped certain kinds of plans can proceed? (Maybe not safe because if a role was dropped, we shouldn't, say, commit any DDLs or writes.) It's also not completely clear whether we want `Command::Commit` to always succeed. It's not unreasonable for that to fail, and then it's up to the AdapterClient to decide what to do next. I'm going to investigate the pgwire vs http differences first. We kinda made up the http protocol and it roughly matches pgwire, but perhaps not exactly in this scenario.

**Comment by antiguru** [comment]
Fix in #28506.
